plugin/pkg/admission/resourcequota/controller.go

```diff
@@ -0,0 +1,477 @@
+	for i := range quotas {
+		resourceQuota := quotas[i]
+		match := evaluator.Matches(&resourceQuota, inputObject)
+		if !match {
+			continue
+		}
+
+		hardResources := quota.ResourceNames(resourceQuota.Status.Hard)
+		evaluatorResources := evaluator.MatchesResources()
+		requiredResources := quota.Intersection(hardResources, evaluatorResources)
+		err := evaluator.Constraints(requiredResources, inputObject)
+		if err != nil {
+			return nil, admission.NewForbidden(a, fmt.Errorf("Failed quota: %s: %v", resourceQuota.Name, err))
+		}
+		if !hasUsageStats(&resourceQuota) {
+			return nil, admission.NewForbidden(a, fmt.Errorf("Status unknown for quota: %s", resourceQuota.Name))
+		}
+
+		interestingQuotaIndexes = append(interestingQuotaIndexes, i)
+	}
+	if len(interestingQuotaIndexes) == 0 {
+		return quotas, nil
+	}
+
+	// Usage of some resources cannot be counted in isolation. For example when
+	// the resource represents a number of unique references to external
+	// resource. In such a case an evaluator needs to process other objects in
+	// the same namespace which needs to be known.
+	if accessor, err := meta.Accessor(inputObject); namespace != "" && err == nil {
+		if accessor.GetNamespace() == "" {
+			accessor.SetNamespace(namespace)
+		}
+	}
+
+	// there is at least one quota that definitely matches our object
+	// as a result, we need to measure the usage of this object for quota
+	// on updates, we need to subtract the previous measured usage
+	// if usage shows no change, just return since it has no impact on quota
+	deltaUsage := evaluator.Usage(inputObject)
+	if admission.Update == op {
+		prevItem, err := evaluator.Get(namespace, name)
+		if err != nil {
+			return nil, admission.NewForbidden(a, fmt.Errorf("Unable to get previous: %v", err))
+		}
+		prevUsage := evaluator.Usage(prevItem)
+		deltaUsage = quota.Subtract(deltaUsage, prevUsage)
+	}
+	if quota.IsZero(deltaUsage) {
+		return quotas, nil
+	}
+
+	for _, index := range interestingQuotaIndexes {
+		resourceQuota := quotas[index]
+
+		hardResources := quota.ResourceNames(resourceQuota.Status.Hard)
+		requestedUsage := quota.Mask(deltaUsage, hardResources)
+		newUsage := quota.Add(resourceQuota.Status.Used, requestedUsage)
+		if allowed, exceeded := quota.LessThanOrEqual(newUsage, resourceQuota.Status.Hard); !allowed {
+			failedRequestedUsage := quota.Mask(requestedUsage, exceeded)
+			failedUsed := quota.Mask(resourceQuota.Status.Used, exceeded)
+			failedHard := quota.Mask(resourceQuota.Status.Hard, exceeded)
+			return nil, admission.NewForbidden(a,
+				fmt.Errorf("Exceeded quota: %s, requested: %s, used: %s, limited: %s",
+					resourceQuota.Name,
+					prettyPrint(failedRequestedUsage),
+					prettyPrint(failedUsed),
+					prettyPrint(failedHard)))
+		}
+
+		// update to the new usage number
+		quotas[index].Status.Used = newUsage
+	}
+
+	return quotas, nil
+}
+
+func (e *quotaEvaluator) evaluate(a admission.Attributes) error {
+	waiter := newAdmissionWaiter(a)
```

Comment:
can we do more checkers before we add the work to the waiter?

for example, it seems like we could check the following safely before paying the cost of the work queue:

1. is this an operation quota even cares about, if not return now
2. are there quotas in the namespace?, if not, return now
3. are the quotas matching on the resource? if not, return now

This would ensure that all requests are not paying the cost of the queue waking every 1s.